src/hotspot/share/runtime/runtimeUpcalls.cpp

```diff
@@ -0,0 +1,210 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/methodDetails.hpp"
+#include "runtime/runtimeUpcalls.hpp"
+
+//--------------------------- initialization -------------------------------------
+GrowableArray<RuntimeUpcallInfo*>* RuntimeUpcalls::_upcalls[RuntimeUpcallType::numTypes];
+RuntimeUpcalls::State RuntimeUpcalls::_state = RuntimeUpcalls::Uninitialized;
+
+bool runtimeUpcalls_open_registration() {
+  return RuntimeUpcalls::open_upcall_registration();
+}
+
+bool RuntimeUpcalls::open_upcall_registration() {
+  assert(_state == Uninitialized, "upcalls are already open");
+  for (int i = 0; i < RuntimeUpcallType::numTypes; i++) {
+    RuntimeUpcalls::_upcalls[i] = nullptr;
+  }
+  _state = Open;
+  return true;
+}
+
+void runtimeUpcalls_close_registration() {
+  RuntimeUpcalls::close_upcall_registration();
+}
+
+void RuntimeUpcalls::close_upcall_registration() {
+  assert(_state == Open, "upcalls are not open");
+  _state = Closed;
+}
+
+//--------------------------- private -------------------------------------
+void RuntimeUpcalls::mark_for_upcalls(RuntimeUpcallType upcallType, const methodHandle& method) {
+  if (_upcalls[upcallType] != nullptr) {
+    MethodDetails md(method);
+    for(RuntimeUpcallInfo* info : *(_upcalls[upcallType])) {
+      if(info->includes(md)) {
+        switch(upcallType) {
+          case onMethodEntry: method->set_has_upcall_on_method_entry(true); break;
+          case onMethodExit:  method->set_has_upcall_on_method_exit(true);  break;
+          default:            ShouldNotReachHere();
+        }
+        break;
+      }
+    }
+  }
+}
+
+bool RuntimeUpcalls::register_upcall(RuntimeUpcallType upcallType, RuntimeUpcallInfo* info) {
+  assert(info != nullptr, "upcall info is null");
+  if (_upcalls[upcallType] == nullptr) {
+    _upcalls[upcallType] = new (mtServiceability) GrowableArray<RuntimeUpcallInfo*>(1, mtServiceability);
+  }
+  info->set_index(_upcalls[upcallType]->length());
+  _upcalls[upcallType]->append(info);
+  return true;
+}
+
+int RuntimeUpcalls::get_num_upcalls(RuntimeUpcallType upcallType) {
+  return (_upcalls[upcallType] == nullptr) ? 0 : _upcalls[upcallType]->length();
+}
+
+void RuntimeUpcalls::upcall_redirect(RuntimeUpcallType upcallType, JavaThread* current, Method* method) {
+  MethodDetails md(method);
+
+  // one possible optimization here is to use the method flags bits to indicate which upcalls to call
```

Comment:
nit: please add some periods/capitalization to clarify sentence boundaries in this set of comments